operations/operations.py

```diff
@@ -0,0 +1,185 @@
+    If surrounding pixels have a mix of True/False the val is left as-is
+
+    This function handles border pixels as if they aren't there. So that
+    the following happens:
+    ------------------------
+    |F|T|     ->      |T|T|
+    |T|T|             |T|T|
+    -----------------------
+
+    Parameters
+    ----------
+    mask: Existing mask with some positions masked
+
+    Returns
+    -------
+    mask: Mask copy after completing the op. described above
+
+    """
+
+    mask = mask.copy()
+
+    def make_flip(fill):
+        flip = sc.Variable(dims=['neighbor', 'y', 'x'],
+                           shape=[
+                               8,
+                           ] + mask.shape,
+                           dtype=sc.dtype.bool)
+        flip['neighbor', 0] = _shift(mask, "x", True, fill)
+        flip['neighbor', 1] = _shift(mask, "x", False, fill)
+        flip['neighbor', 2] = _shift(mask, "y", True, fill)
+        flip['neighbor', 3] = _shift(mask, "y", False, fill)
+        flip['neighbor', 4:6] = _shift(flip['neighbor', 0:2], "y", True, fill)
+        flip['neighbor', 6:8] = _shift(flip['neighbor', 0:2], "y", False, fill)
+        return flip
+
+    # mask if all neighbors masked
+    mask = mask | sc.all(make_flip(True), 'neighbor')
+    # unmask if no neighbor masked
+    mask = mask & sc.any(make_flip(False), 'neighbor')
+    return mask
+
+
+def mean_from_adj_pixels(data):
+    """
+    Applies a mean across 8 neighboring pixels (plus centre value)
+    for data with 'x' and 'y' dimensions (at least).
+    Result will calculate mean from slices across additional dimensions.
+    
+    For example if there is a tof dimension in addition to x, and y, 
+    for each set of neighbours the returned mean will take the mean tof value in the neighbour group.
+    """
+    fill = np.finfo(data.values.dtype).min
+    has_variances = not data.variances is None
+    container = sc.Variable(['neighbor'] + data.dims,
+                            shape=[
+                                9,
+                            ] + data.shape,
+                            variances=has_variances)
+    container['neighbor', 0] = data
+    container['neighbor', 1] = _shift(data, "x", True, fill)
+    container['neighbor', 2] = _shift(data, "x", False, fill)
+    container['neighbor', 3] = _shift(data, "y", True, fill)
+    container['neighbor', 4] = _shift(data, "y", False, fill)
+    container['neighbor', 5:7] = _shift(container['neighbor', 1:3], "y", True,
+                                        fill)
+    container['neighbor', 7:9] = _shift(container['neighbor', 1:3], "y", False,
+                                        fill)
+
+    edges_mask = sc.less_equal(container, sc.Variable(value=fill))
+    da = sc.DataArray(data=container, masks={'edges': edges_mask})
+    return sc.mean(da, dim='neighbor').data
+
+
+def _median(neighbors, edges_mask, dim):
+    masked_values = np.ma.array(neighbors.values,
```

Comment:
Interesting to see numpy masked arrays interacting with scipp masks.

Reply:
There may be further possibilities to elide allocations, such as in-place for the `ma.median` though I don't know if that will have any large effect on performance given that we do a 9 * larger allocation anyway for the neighbours.